File: quanta_tissu/tisslm/program/layers/layernorm.cpp
#include "layernorm.h"
#include <cmath>

LayerNorm::LayerNorm(int d_model, float eps)
    : gamma(1, d_model), beta(1, d_model), d_gamma(1, d_model), d_beta(1, d_model), eps(eps) {
    for (int i = 0; i < d_model; ++i) {
        gamma({0, (size_t)i}) = 1.0f;
        beta({0, (size_t)i}) = 0.0f;
    }
}

Matrix LayerNorm::forward(const Matrix& x) {
    size_t rows = x.rows(), cols = x.cols();
    Matrix res(rows, cols);
    for (size_t i = 0; i < rows; ++i) {
        float m = 0.0f; for (size_t j = 0; j < cols; ++j) m += x({i, j});
        m /= cols;
        float v = 0.0f; for (size_t j = 0; j < cols; ++j) v += std::pow(x({i, j}) - m, 2);
        v /= cols;
        float inv_std = 1.0f / std::sqrt(v + eps);
        for (size_t j = 0; j < cols; ++j) res({i, j}) = (x({i, j}) - m) * inv_std * gamma({0, j}) + beta({0, j});
    }
    return res;
}

Matrix LayerNorm::backward(const Matrix& d_out, const Matrix& x) {
    size_t rows = x.rows(), cols = x.cols();
    Matrix dx(rows, cols);
    for (size_t i = 0; i < rows; ++i) {
        float m = 0.0f; for (size_t j = 0; j < cols; ++j) m += x({i, j});
        m /= cols;
        float v = 0.0f; for (size_t j = 0; j < cols; ++j) v += std::pow(x({i, j}) - m, 2);
        v /= cols;
        float inv_std = 1.0f / std::sqrt(v + eps);
        float d_m = 0.0f, d_v = 0.0f;
        for (size_t j = 0; j < cols; ++j) {
            float x_hat = (x({i, j}) - m) * inv_std;
            d_gamma({0, j}) += d_out({i, j}) * x_hat;
            d_beta({0, j}) += d_out({i, j});
            float d_x_hat = d_out({i, j}) * gamma({0, j});
            d_v += d_x_hat * (x({i, j}) - m) * (-0.5f) * std::pow(v + eps, -1.5f);
            d_m += d_x_hat * (-inv_std);
        }
        for (size_t j = 0; j < cols; ++j) {
            dx({i, j}) = d_out({i, j}) * gamma({0, j}) * inv_std + d_v * 2.0f * (x({i, j}) - m) / cols + d_m / cols;
        }
    }
    return dx;
}

File: quanta_tissu/tisslm/js/transformer/LayerNorm.js
const { Matrix } = require('../core/Matrix.js');

class LayerNorm {
    constructor(features, epsilon = 1e-5) {
        this.epsilon = epsilon;
        this.gamma = Matrix.ones(1, features); // Learnable scale parameter
        this.beta = Matrix.zeros(1, features); // Learnable shift parameter
    }

    forward(x) {
        // x is a Matrix with shape (batch_size, features)
        const mean = x.data.map(row => row.reduce((a, b) => a + b, 0) / row.length);
        const variance = x.data.map((row, i) => {
            return row.reduce((acc, val) => acc + Math.pow(val - mean[i], 2), 0) / row.length;
        });

        const normalizedData = x.data.map((row, i) => {
            const std = Math.sqrt(variance[i] + this.epsilon);
            return row.map(val => (val - mean[i]) / std);
        });

        const normalizedMatrix = new Matrix(normalizedData);

        // Apply scale and shift
        const scaled = normalizedMatrix.multiply(this.gamma); // Element-wise
        const shifted = scaled.add(this.beta); // Element-wise

        return shifted;
    }
}

// Add a static ones method to Matrix if it's missing from the original implementation
if (!Matrix.ones) {
    Matrix.ones = function(rows, cols) {
        const data = Array(rows).fill(0).map(() => Array(cols).fill(1));
        return new Matrix(data);
    }
}


module.exports = { LayerNorm };
